src/utils.jl

```diff
@@ -0,0 +1,156 @@
+    size(V, dim) == size(M, 1) || throw(DomainError(
+        "Size of V along dim must be same as size of M"))
+    size(out) == size(V) || throw(DomainError("Input and output arrays must have same size"))
+    dim == 1 || throw(DomainError("Cannot multiply along dimension $dim of 1-D array"))
+    mul!(out, M, V)
+end
+
+function dot!(out, M, V::AbstractArray{T, 2}; dim=1) where T
+    size(V, dim) == size(M, 1) || throw(DomainError(
+        "Size of V along dim must be same as size of M"))
+    size(out) == size(V) || throw(DomainError("Input and output arrays must have same size"))
+    dim <= 2 || throw(DomainError("Cannot multiply along dimension $dim of 2-D array"))
+    if dim == 1
+        mul!(out, M, V)
+    else
+        Vtmp = permutedims(V, (2, 1))
+        tmp = M * Vtmp
+        permutedims!(out, tmp, (2, 1))
+    end
+end
+
+function dot!(out, M, V::AbstractArray; dim=1) where T
+    size(V, dim) == size(M, 1) || throw(DomainError(
+        "Size of V along dim must be same as size of M"))
+    size(out) == size(V) || throw(DomainError("Input and output arrays must have same size"))
+    dim <= ndims(V) || throw(DomainError(
+        "Cannot multiply along dimension $dim of $(ndims(V))-D array"))
+    if dim == 1
+        idxhi = CartesianIndices(size(V)[3:end])
+        _dot!(out, M, V, idxhi)
+    else
+        dims = collect(1:ndims(V)) # [1, 2, ..., ndims(V)]
+        otherdims = filter(d -> d ≠ dim, dims) # dims but without working dimension
+        #= sort dimensions by their size so largest other dimension is part of matrix-matrix
+            multiplication, for speed. other dimensions are iterated over in _dot! =#
+        sidcs = sortperm(collect(size(V)[otherdims]), rev=true)
+        perm = (dim, otherdims[sidcs]...)
+        iperm = invperm(perm) # permutation to get back to original size
+        Vtmp = permutedims(V, perm)
+        tmp = similar(Vtmp)
+        idxhi = CartesianIndices(size(Vtmp)[3:end])
+        _dot!(tmp, M, Vtmp, idxhi)
+        permutedims!(out, tmp, iperm)
+    end
+end
+
+function _dot!(out, M, V, idxhi)
+    for hi in idxhi
+        mul!(view(out, :, :, hi), M, view(V, :, :, hi))
+    end
+end
+
+"""
+    dimdot(v, A; dim=1)
+
+Calculate the dot product between vector `v` and one dimension of array `A`, iterating over
+all other dimensions.
+"""
+function dimdot(v, A; dim=1)
+    dims = collect(size(A))
+    dims[dim] = 1
+    out = Array{eltype(A)}(undef, Tuple(dims))
+    dimdot!(out, v, A; dim=dim)
+    return out
+end
+
+function dimdot(v, A::Vector; dim=1)
+    return dot(v, A)
+end
+
+function dimdot!(out, v, A; dim=1)
+    idxlo = CartesianIndices(size(A)[1:dim-1])
+    idxhi = CartesianIndices(size(A)[dim+1:end])
+    _dimdot!(out, v, A, idxlo, idxhi)
+end
+
+function _dimdot!(out, v, A, idxlo, idxhi)
+    for lo in idxlo
+        for hi in idxhi
+            out[lo, 1, hi] = dot(v, view(A, lo, :, hi))
+        end
+    end
+end
+
+function sphbesselj_scale(n::Integer)
```

Comment:
Since `ceven` is constant (with constant type), couldn't this be much simpler?
```suggestion
sphbesselj_scale(n::Integer) = isodd(n) ? 1.0 : √(π / 2)
```

Reply:
Yes that is simpler and should be type stable.